### PR TITLE
DrivAerML: EMA=0.9995 + gc=0.3 (softer gc on champion)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The DrivAerML champion (PR #3072, eren) uses gc=0.5 with EMA=0.9995, achieving 3.833% val_primary/surface_rel_l2_pct. On TandemFoil, softer grad clipping showed monotonic improvement: gc=1.0 → 0.5 → 0.3 (PR #3108, zenitsu). gc=0.25 was tested on DrivAerML by gilbert (#3151 — currently in-flight).

This PR fills the gap in the DrivAerML gc sweep: **0.25, 0.3, 0.5, 1.0**. If the TandemFoil pattern transfers, gc=0.3 could beat the gc=0.5 champion.

Key stability question: gc=0.5 was specifically identified as the stability guard that makes EMA viable on DrivAerML (EMA alone diverged at ep28 without gc). Does gc=0.3 remain stable through cosine restarts, or does it destabilize like gc=1.0? We expect it to survive based on the TandemFoil precedent, but this needs empirical confirmation.

## Instructions

Run the following training command. All flags are identical to the current champion (PR #3072) except `--grad-clip` is changed from 0.5 to 0.3:

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --grad-clip 0.3 \
  --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 --wandb-group dm-ema-gc-sweep
```

**What to report:**
- Best `val_primary/surface_rel_l2_pct` and the epoch it was achieved
- Stability: does training remain stable through cosine restarts, or does it diverge? (Watch for any spike at the start of a new cosine cycle)
- Comparison to gc=0.5 champion (baseline: 3.833%)
- W&B run ID

The `--wandb-group dm-ema-gc-sweep` groups this run with the gc=0.25 (gilbert #3151), gc=0.5 (eren #3072), and any gc=1.0 runs for easy sweep comparison.

## Baseline

| Metric | Value |
|--------|-------|
| val_primary/surface_rel_l2_pct | **3.833%** |
| epoch | 511 |
| test/surface_rel_l2_pct | 4.685% |

- Best PR: #3072 (eren — `eren/dm-ema-9995-gc05`)
- W&B run: `ncl1dh88`
- Config: 4L/512d/8H, AdamW lr=5e-4, T_max=30, gc=0.5, EMA=0.9995, Fourier features
- External target: <3.71% (AB-UPT) — gap to close: 0.123 pp

Reproduce champion:
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --grad-clip 0.5 \
  --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT himmel:

## Results

### Experiment: DrivAerML gc=0.3 with EMA=0.9995

| Metric | Baseline (gc=0.5) | gc=0.3 |
|--------|------|------|
| **Best val surface_rel_l2_pct** | **3.833%** (ep511) | **8.816%** (ep80) |
| Stable through epoch | 511+ | 80 |
| First gradient explosion | None reported | ep81 (grad_norm=17.8) |
| Terminal divergence | N/A | ep82-90 (grad_norm 113-430) |

**External references:** Transolver-3 3.71% | AB-UPT 3.82%

### Command
```bash
SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --grad-clip 0.3 \
  --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 --wandb-group dm-ema-gc-sweep \
  --agent himmel --wandb-name "himmel/dm-ema-gc03"
```

### W&B Run ID
`og8ny8s7`

### Peak memory
Single H100 (98GB). No OOM issues.

### Bug fix
Found and fixed a `primary_metric_key` shadowing bug in train.py (line 1640-1643). The local variable assignment `primary_metric_key = "val_primary/..."` shadowed the function `primary_metric_key()` defined at line 1423, causing `UnboundLocalError` at line 1634. Fix: removed the shadowing assignment and reused `best_val_primary_metric_name` (which already calls the function). This bug was present on the `radford` branch and would have affected any new training run.

### What happened

**Hypothesis falsified.** gc=0.3 is not stable enough for DrivAerML 4L/512d. The TandemFoil gc monotonic improvement pattern does NOT transfer.

**Training trajectory:**
- **Epochs 1-66:** Perfect convergence. Best improved nearly every epoch (10.20% at ep66). All cosine restarts handled cleanly. Grad norms declined steadily from 9.5→0.71. Clip events gradually decreased (394→362/394), suggesting gradients were settling below the 0.3 threshold.
- **Epochs 67-72:** First regression dip. Val temporarily worsened to 11.97% but grad norms stayed healthy (0.66-0.78). This was a normal EMA adjustment phase — the underlying model was still learning (loss continued decreasing).
- **Epochs 73-80:** Recovery and new bests. Val improved from 11.97% → 8.82% (best at ep80). This appeared to be the strongest phase of training.
- **Epoch 81:** First gradient spike (grad_norm=17.8). Val only mildly affected (9.11%).
- **Epoch 82-90: Terminal divergence.** Grad norms exploded to 113→288→430, val deteriorated to 44-46%. Brief partial recovery attempts (grad_norm dropped to 99 at ep87) were followed by re-explosions. Run killed at epoch 90.

**Why gc=0.3 fails but gc=0.5 doesn't:** The explosion occurred at epoch 81 during a cosine cycle, not at a restart boundary — so it's not the restart itself that's dangerous. Rather, gc=0.3 allows larger per-batch updates (max_norm=0.3 vs 0.5 clips less aggressively), which accumulates more update magnitude over many steps. When the model enters a region with steep loss curvature (common in transformer training around epoch 80+ on DrivAerML), the larger effective step size from softer clipping tips the model past a stability threshold. gc=0.5 provides enough damping to navigate these regions.

**The 8.82% best is notably worse than the 3.833% baseline** — the run diverged 431 epochs before the champion even hit its best. gc=0.3 both destabilizes earlier AND never reaches competitive performance during its stable phase.

### Suggested follow-ups
1. **gc=0.4 sweep point:** If gc=0.3 diverges at ep81 and gc=0.5 is stable through 511+, gc=0.4 might find the sweet spot — softer than champion but still stable.
2. **gc=0.3 + warmup clipping:** Start with gc=0.5 for the first 100 epochs, then relax to gc=0.3. The model may be more robust to softer clipping once past the initial instability-prone phase.
3. **Adaptive gradient clipping (AGC):** Per-layer clipping by ||g||/||W|| ratio (~0.01) would be a principled alternative to hand-tuning scalar thresholds. Literature (Brock et al., ICML 2021) supports this for transformers.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3162 because ADVISOR: gc=0.3 diverges catastrophically at ep81 (grad norm 17.8→430), best 8.816% — 2.3x worse than 3.833% baseline. gc=0.5 is now quad-confirmed as the sharp optimum for DM EMA regime (0.25=13.1% starves, 0.3=8.8% diverges, 0.5=3.833% champion, 1.0=6.21% insufficient). No further gc sweeps below 0.5 warranted. Your AGC suggestion is interesting — assigning that as a fresh experiment. The bug fix (primary_metric_key shadowing) should be submitted as a separate bugfix PR if it's a real issue..
